README.md
## Yawls (Yet Another Webcam Light Sensor):
Adjust the brightness level of your display by using the internal/external camera of your Notebook as an ambient light sensor and is designed to combine comfort and energy saving. Your privacy is a high good and therefore all images are processed internally of Yawls. So that **no image** is saved on your drive or is send to someone in the internet [`1`][1] [`2`][2].

![screenshot](http://i.imgur.com/rxrJgLs.png)

If you install Yawls it runs as a system daemon in background in an interval of 30 seconds per default.
Therefore other programs are also able  to use the camera. The interval is divided into four configurable
variables which will be explained later. You are also still able to correct the screen brightness with your function keys and other applets because Yawls won't update the brightness if your brightness condtions are stable. 

If you have Yawls fresh installed [please enable the universe packages in Ubuntu to avoid dependency errors] it is handy to calibrate it because otherwise it may not function properly.
There are three ways to calibrate Yawls:

* via CLI: run once in a dark and once in a bright room [yawls -c] 
* via GUI: there is a calibration wizard under preferences **(recommended)**
* Yawls is also able to calibrate itself over time.

Since version 1.2.0 Yawls is also able to detect backlit conditions to avoid a low screen brightness in such situations.
A good example for such a situation is if you sit with your back to a great window or the other way around.

As mentioned above Yawls has some configurable parameters which you can edit with the GUI or direct in the configuration file (/etc/yawls.properties). If you use the GUI there are tooltips when you hover with your mouse cursor over the parameter name.

![tooltip gif](http://i.imgur.com/GsaCOVt.gif)

### Parameter description

##### lowerLimit/upperLimit/minIncrease/maxIncrease:
The lowerLimit defines the shortest time in milliseconds between the camera activations to check the ambient light conditions. This value will be increased over time with the value of maxIncrease
if the light conditions are stable. This will be go on until it reaches the upperLimit.
If there is a change in the ambient light condition Yawls will fall back to the lowerLimit for a short time and jumps back to the value where it came from if the light conditions are stable.
Yawls will also increase the lowerLimit if your Notebook battery falls under 30% to be more battery friendly.

Default values:
- lowerLimit: 500
- upperLimit: 30000
- minIncrease: 100
- maxIncrease: 500

##### faceDetect:
If set to true Yawls tries to detect if someone is or is not in front of the Notebook. If Yawls does not detect any faces with the camera the screen will be dimmed to save battery (here Yawls fall
back to the lowerLimit and use the minIncrease like described above). If Yawls detect a face or you came back to your Notebook the screen wakes up and Yawls works like normal.

Default: false

##### darkeningThreshold:
This variable controls the darkening threshold. Here an example which is also included in the configuration file a value of 10 means that the current brightness has to be less than 75% of the last measured
ambient brightness. Only values between 0 and 100 are allowed.

Default: 50

##### brighteningThreshold:
This variable controls the brightening threshold. Here an example which is also included in the configuration file a value of 10 means that the current brightness has to be greater than 110 percent of the
last measured brightness.

Default: 25

##### logLevel:
The logLevel defines how detailed the log is: 0 (error) < 1 (warning) < 2 (info) < 3 (debug).

Default: 1

### Dependencies
* openjdk-7-jre-headless
* gettext-base
* libgtk-3-0
* libjava-gnome-java
* libopencv2.4-java
* policykit-1-gnome

### Issue tracking
[Report a bug on launchpad.net](https://bugs.launchpad.net/yawls)

### Help to translate
[Translations on Launchpad.net](https://translations.launchpad.net/yawls)

### Get in touch
* [Twitter](https://twitter.com/TheDsWeb)
* [German project blog](http://thedsweb.blogspot.de/p/yawls.html)

### Download of the current stable release and PPA
[yawls_1.2.2.deb](https://launchpad.net/yawls/1.2.x/1.2.2/+download/yawls_1.2.2_all.deb)

https://launchpad.net/~thedsweb/+archive/ubuntu/yawls-daily

### Build dependencies
* javahelper
* openjdk-7-jdk
* gettext-base
* libjava-gnome-java
* libopencv2.4-java
* ant
* po4a

[1]: https://github.com/TheDsWeb/yawls/blob/master/src/com/blogspot/thedsweb/engine/Brightness.java#L166-L198
[2]: https://github.com/TheDsWeb/yawls/blob/master/src/com/blogspot/thedsweb/engine/Face.java#L33-L64


debian/changelog
yawls (1.2.2) stable; urgency=low

  * New upstream release
  * Update GUI
    - Split GUI windows into separate classes
    - Update configuration entries after reset
  * Update yawls_cli.sh
    - Add better SIGTERM/SIGINT support (LP: #1475841)
  * Update engine/Brightness.java
    - Improved performance
  * Update translations
  * Add backlight device interface
    - Make it possible to add other backlight control methods.
  * Add engine/MinMaxContainer.java
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sun, 06 Sep 2015 14:32:16 +0100

yawls (1.2.1) stable; urgency=low

  * New upstream release
  * Update bulid.sh
    - Merge GUI and CLI versions
  * Update build.xml
    - To prevent some build errors
  * Update gui/Show.java
    - Revamped GUI (LP: #1433997)
    - Fix wrong logrotate config (LP: #1470463)
    - Fix log level is not saved via GUI
  * Update util/SingleInstance.java
    - Rename to LockFile.java
    - Can now handle more than one lock file
  * Update debian/yawls.init
    - iosched and procsched removed
  * Add util/ReadParallel.java and util/WriteParallel.java
    - Read and write with worker processes
  * Add engine/Fade.java
    - Smoother change between brightness level
  * Add gui/ValueDaemon.java
    - Update camera stats every second
  * Add main/Debug.java
    - Create logfile for debugging
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 17 Jul 2015 13:05:54 +0100

yawls (1.1.2) stable; urgency=low

  * New upstream release
  * Update src/*
  	- Add license header to all Java files
  	- Improve Java Convention
  * Add policykit file
  	- Move from gksudo to pkexec
  	- Update yawls_gui.sh
  * Update debian/control
  	- Add version range for libgtk-3-0 (3.10 < 3.14)
  * Upadte main/Yawls.java
  	- Add daemon/BrightnessDaemon.java, daemon/Tools.java
  * Update engine/Brightness.java
  	- Improved performance
  * Update engine/Face.java
  	- Improved performance
  * Update error/IllegalArgumentINConfig.java
  	- Change name to IllegalArgumentINConfigException
  * Upadte error/NoBacklightDeviceFound.java
  	- Change name to NoBacklightDeviceFoundException
  * Update yawls_cli.sh
  	- Improved memory footprint of yawls Java instance
  	- Fix PID file error if not run as root (LP: #1424250)
  	- Fix PID file doesn't save program ID (LP: #1425129)
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sun, 22 Mar 2015 15:31:16 +0100

yawls (1.1.1) stable; urgency=low

  * New upstream release
  * Update build.xml
  	- Use reference for libopencv
  * Update debian/control
  	- Add libopencv2.4-java as dependency
  * Update yawls_cli.sh
  	- Add workaround for broken opencv symlinks on some systems
  * Update debian/yawls-cli.links
  * Update debian/yawls-gtk.links
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sat, 21 Feb 2015 16:25:00 +0100
 
yawls (1.1.0) stable; urgency=low

  * New upstream release
  * Add gui/Show.java, util/Check.java
  	- Make config editable on-the-fly
  	- Add calibration wizzard
  * Add yawls_cli.sh, yawls_gui.sh
  	- Make daemon more robust
  	- Add workaround for GTK error
  * Add engine/Stable.java
  	- Add stable light conditions detection function to improve battery impact
  * Add error/IllegalArgumentINConfig.java, error/NoBacklightDeviceFound.java
  * Add Yawls.desktop, yawls_icon_64x64.png
  * Add yawls_about.glade, yawls_calibrate.glade
  * Add yawls_error.glade, yawls_gui.glade
  * Update debian/control
  	- Split into two versions (CLI, GUI)
  * Update debian/rules
  * Update build.xml, build.sh
  	- Make builds more stable
  * Update manual
  	- Manual is now translateable
  	- Add that daemon mode is now default
  * Update util/Translate.java, util/Database.java
  * Update main/Yawls.java
  	- Improved daemon mode
  	- Add -g, --gui options
  * Update engine/Brightness.java
  	- Improve automatic camera calibration
  * Update util/GeneratePath.java
  	- Improve BestType function to make backlight type detection more stable
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 20 Feb 2015 15:04:48 +0100

yawls (1.0.1) stable; urgency=low

  * New upstream release
  * Add OpenCV PPA for precise compatibility
  * Change src/*
  	- Improved Java Convention
  * Update src/engine/Brightness.java
  	- Fix does not internally store the last ambient brightness (#1411946)
  	- Fix screen dim while camera is used by another program (#1409247)
  	- Fix Problem with auto-calibration (#1416639)
  	- Change reference value type to double to improve reference calculation
  * Update src/engine/Face.java
  	- Optimze face detection
  * Update src/util/Database.java, src/util/Config.java
  	- Change path of config file
  * Update src/util/GeneratePath.java
  	- Add bestType detection for GeneratePath
  * Update src/util/Translation.java
  	- Add gettext library for easier translations and translation integration
  	- Update build.sh
  	- Update build.xml
  * Add src/engine/Threshold.java
  	- Add Threshold to config for brightening and darkening
  * Update debian/compat
  	- Bump Version to 9
  * Update debian/rules
  	- Add override for update-rc to install init script without activation
  * Update debian/control
  	- Bump debhelper Version to >= 9
  	- Add Homepage URL
  	- Add Version Control System URLs
  * Update debian/yawls.install
  	- Remove unneeded path
  * Add debian/yawls.postrm.debhelper and debian/yawls.prerm.debhelper
  	- For a cleaner uninstall if daemon is added to defaults by the user
  * Update debian/changelog
  	- Changelog is now more detailed
  * Update man/yawls.*
  	- Path correction of config files
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 23 Jan 2015 17:00:28 +0100

yawls (1.0.0) stable; urgency=low

  * New upstream release
  	- Clean up dependencies
  	- Add possibility to translate the program
  	- Add translated man pages
  	- Add translation for german (program and man page)
  	- Add config file for daemon (/etc/default/yawls.properties)
  		- Add option to detect Faces

src/com/blogspot/thedsweb/gui/MainGUI.java
/*******************************************************************************
 * Copyright (c) 2015 Dominik Brämer.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/
package com.blogspot.thedsweb.gui;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.util.MissingResourceException;
import java.util.Properties;
import java.util.logging.Level;

import org.gnome.gdk.Event;
import org.gnome.gdk.Pixbuf;
import org.gnome.gtk.Align;
import org.gnome.gtk.Button;
import org.gnome.gtk.Frame;
import org.gnome.gtk.Gtk;
import org.gnome.gtk.HBox;
import org.gnome.gtk.HButtonBox;
import org.gnome.gtk.Label;
import org.gnome.gtk.Menu;
import org.gnome.gtk.MenuBar;
import org.gnome.gtk.MenuItem;
import org.gnome.gtk.MenuItem.Activate;
import org.gnome.gtk.SpinButton;
import org.gnome.gtk.Statusbar;
import org.gnome.gtk.Switch;
import org.gnome.gtk.VBox;
import org.gnome.gtk.Widget;
import org.gnome.gtk.Window;
import org.gnome.gtk.Window.DeleteEvent;
import org.gnome.gtk.WindowPosition;

import com.blogspot.thedsweb.main.Debug;
import com.blogspot.thedsweb.util.Check;
import com.blogspot.thedsweb.util.Config;
import com.blogspot.thedsweb.util.Database;
import com.blogspot.thedsweb.util.Files;
import com.blogspot.thedsweb.util.LockFile;
import com.blogspot.thedsweb.util.Translate;

public class MainGUI {
    private static Window mainGUI;
    private static Statusbar status;
    private static Switch fdSwitch;
    private static SpinButton llSpin;
    private static SpinButton ulSpin;
    private static SpinButton miInSpin;
    private static SpinButton maInSpin;
    private static SpinButton dtSpin;
    private static SpinButton btSpin;
    private static SpinButton loSpin;
    private static Label minValue;
    private static Label maxValue;
    private static Label lastValue;

    public MainGUI() {
	mainGUI = new Window();
	mainGUI.setSizeRequest(550, 0);
	mainGUI.setResizable(false);
	mainGUI.setTitle("Yawls");
	mainGUI.setPosition(WindowPosition.CENTER);
	mainGUI.setBorderWidth(5);
	Pixbuf icon;
	try {
	    icon = new Pixbuf(Database.PATH_ICON);
	    mainGUI.setIcon(icon);
	} catch (final FileNotFoundException e) {
	    Debug.LOG.log(Level.WARNING, "Could not find yawls icon.");
	}

	status = new Statusbar();

	final VBox ground = new VBox(false, 4);

	// Create MenuBar
	final MenuBar menuBar = new MenuBar();
	constructMenu(menuBar);

	final HBox cage = new HBox(false, 2);

	final Frame frame1 = new Frame(null);
	final Label stats = new Label("<b>" + Translate._("Webcam stats:")
		+ "</b>");
	stats.setUseMarkup(true);
	frame1.setLabelWidget(stats);
	constructWebcamStats(frame1);
	frame1.setBorderWidth(1);
	cage.add(frame1);

	final Frame frame2 = new Frame(null);
	final Label conf = new Label("<b>" + Translate._("Configuration:")
		+ "</b>");
	conf.setUseMarkup(true);
	frame2.setLabelWidget(conf);
	constructConfiguration(frame2);
	frame2.setBorderWidth(1);
	cage.add(frame2);

	final HButtonBox buttonBox = new HButtonBox();
	constructButton(buttonBox);

	status.setMessage(checkStatus());

	ground.packStart(menuBar, false, false, 0);
	ground.packStart(cage, false, false, 0);
	ground.packStart(buttonBox, false, false, 0);
	ground.packEnd(status, false, false, 0);

	mainGUI.add(ground);

	mainGUI.connect(new DeleteEvent() {
	    @Override
	    public boolean onDeleteEvent(Widget arg0, Event arg1) {
		Gtk.mainQuit();
		return false;
	    }
	});
    }

    public void start() {
	mainGUI.showAll();
    }

    private String checkStatus() {
	String status = "";
	if (!Check.backlightDevice()) {
	    status = Translate._("No backlight device found");
	    if (!Check.camera()) {
		status += " ";
		return status + Translate._("and no camera found.");
	    }
	    status += ".";
	} else if (!Check.camera()) {
	    return Translate._("No camera found.");
	}
	return status;
    }

    private void constructButton(HButtonBox buttonBox) {
	final Button reset = new Button();
	reset.setLabel(Translate._("Reset"));
	reset.connect(new Button.Clicked() {
	    @Override
	    public void onClicked(Button arg0) {
		Files.writeConfig();
		readConfig();
	    }
	});
	buttonBox.add(reset);
	final Button apply = new Button();
	apply.setLabel(Translate._("Apply"));
	apply.connect(new Button.Clicked() {
	    @Override
	    public void onClicked(Button arg0) {
		final boolean fd = fdSwitch.isActive();
		final int ll = (int) llSpin.getValue();
		final int ul = (int) ulSpin.getValue();
		final int mi = (int) miInSpin.getValue();
		final int ma = (int) maInSpin.getValue();
		final int dt = (int) dtSpin.getValue();
		final int bt = (int) btSpin.getValue();
		final int lo = (int) loSpin.getValue();

		if (ul <= ll) {
		    status.setMessage(Translate
			    ._("The value of the upper limit must be greater than the lower limit."));
		    return;
		}
		if (mi > ma) {
		    status.setMessage(Translate
			    ._("The value of the max. increase must be greater than the min. increase."));
		    return;
		}
		status.setMessage(checkStatus());

		final Properties config = new Properties();
		config.setProperty("faceDetect", Boolean.toString(fd));
		config.setProperty("lowerLimit", Integer.toString(ll));
		config.setProperty("upperLimit", Integer.toString(ul));
		config.setProperty("minIncrease", Integer.toString(mi));
		config.setProperty("maxIncrease", Integer.toString(ma));
		config.setProperty("darkeningThreshold", Integer.toString(dt));
		config.setProperty("brighteningThreshold", Integer.toString(bt));
		config.setProperty("logLevel", Integer.toString(lo));

		final File cFile = new File(Database.PATH_TO_CONFIG_FILE);
		try {
		    final OutputStream configStream = new FileOutputStream(
			    cFile);
		    config.store(configStream, "Yawls auto generated config");
		} catch (final IOException e) {
		    Debug.LOG.log(Level.WARNING, "Could not create file: "
			    + Database.PATH_TO_CONFIG_FILE, e);
		}

		// Create a lock file to trigger the daemon to
		// reload the configuration on the fly
		LockFile.create(Database.PATH_LOCK_CONFIG, false);
	    }
	});
	buttonBox.add(apply);
	buttonBox.setAlignHorizontal(Align.END);
    }

    private void constructConfiguration(Frame frame) {
	final VBox vbox = new VBox(false, 5);
	vbox.setBorderWidth(10);
	final HBox hbox1 = new HBox(true, 0);
	hbox1.setSizeRequest(275, 0);
	final HBox hbox2 = new HBox(true, 0);
	hbox2.setSizeRequest(275, 0);
	final HBox hbox3 = new HBox(true, 0);
	hbox3.setSizeRequest(275, 0);
	final HBox hbox4 = new HBox(true, 0);
	hbox4.setSizeRequest(275, 0);
	final HBox hbox5 = new HBox(true, 0);
	hbox5.setSizeRequest(275, 0);
	final HBox hbox6 = new HBox(true, 0);
	hbox6.setSizeRequest(275, 0);
	final HBox hbox7 = new HBox(true, 0);
	hbox7.setSizeRequest(275, 0);
	final HBox hbox8 = new HBox(true, 0);
	hbox8.setSizeRequest(275, 0);

	final Label fdLabel = new Label(Translate._("face detect"));
	fdLabel.setAlignHorizontal(Align.START);
	fdLabel.setTooltipText(Translate
		._("Set true to enable or false to disable face detection"));
	final Label llLabel = new Label(Translate._("lower limit"));
	llLabel.setAlignHorizontal(Align.START);
	llLabel.setTooltipText(Translate
		._("Minimum time between camera activations in milliseconds"));
	final Label ulLabel = new Label(Translate._("upper limit"));
	ulLabel.setAlignHorizontal(Align.START);
	ulLabel.setTooltipText(Translate
		._("Maximum time between camera activations in milliseconds"));
	final Label miInLabel = new Label(Translate._("min. increase"));
	miInLabel.setAlignHorizontal(Align.START);
	miInLabel
		.setTooltipText(Translate
			._("Minimum time increase between camera activations in milliseconds if nothing to do"));
	final Label maInLabel = new Label(Translate._("max. increase"));
	maInLabel.setAlignHorizontal(Align.START);
	maInLabel
		.setTooltipText(Translate
			._("Maximum time increase between camera activations in milliseconds if nothing to do"));
	final Label dtLabel = new Label(Translate._("darkening threshold"));
	dtLabel.setAlignHorizontal(Align.START);
	dtLabel.setTooltipText(Translate
		._("Value for darkening threshold in percent (0-100)\nFor example a value of 25 means that the current brightness\nhas to be less than 75 percent of the last measured brightness."));
	final Label btLabel = new Label(Translate._("brightening threshold"));
	btLabel.setAlignHorizontal(Align.START);
	btLabel.setTooltipText(Translate
		._("Value for brightening threshold in percent (0-100)\nFor example a value of 10 means that the current brightness\nhas to be greater than 110 percent of the last measured brightness."));
	final Label loLabel = new Label(Translate._("log level"));
	loLabel.setAlignHorizontal(Align.START);
	loLabel.setTooltipText(Translate
		._("Log level 0 (error) < 1 (warning) < 2 (info) < 3 (debug)"));

	hbox1.add(fdLabel);
	hbox2.add(llLabel);
	hbox3.add(ulLabel);
	hbox4.add(miInLabel);
	hbox5.add(maInLabel);
	hbox6.add(dtLabel);
	hbox7.add(btLabel);
	hbox8.add(loLabel);

	fdSwitch = new Switch();
	fdSwitch.setAlignHorizontal(Align.START);
	llSpin = new SpinButton(500, 600000, 500);
	llSpin.setAlignHorizontal(Align.FILL);
	ulSpin = new SpinButton(1000, 600000, 500);
	ulSpin.setAlignHorizontal(Align.FILL);
	miInSpin = new SpinButton(100, 600000, 100);
	miInSpin.setAlignHorizontal(Align.FILL);
	maInSpin = new SpinButton(100, 600000, 100);
	maInSpin.setAlignHorizontal(Align.FILL);
	dtSpin = new SpinButton(5, 100, 5);
	dtSpin.setAlignHorizontal(Align.FILL);
	btSpin = new SpinButton(5, 100, 5);
	btSpin.setAlignHorizontal(Align.FILL);
	loSpin = new SpinButton(0, 3, 1);
	loSpin.setAlignHorizontal(Align.FILL);

	readConfig();

	hbox1.add(fdSwitch);
	hbox2.add(llSpin);
	hbox3.add(ulSpin);
	hbox4.add(miInSpin);
	hbox5.add(maInSpin);
	hbox6.add(dtSpin);
	hbox7.add(btSpin);
	hbox8.add(loSpin);

	vbox.packStart(hbox1, false, false, 0);
	vbox.packStart(hbox2, false, false, 0);
	vbox.packStart(hbox3, false, false, 0);
	vbox.packStart(hbox4, false, false, 0);
	vbox.packStart(hbox5, false, false, 0);
	vbox.packStart(hbox6, false, false, 0);
	vbox.packStart(hbox7, false, false, 0);
	vbox.packStart(hbox8, false, false, 0);

	frame.add(vbox);
    }

    private void constructWebcamStats(Frame frame) {
	final VBox vbox = new VBox(false, 5);
	vbox.setBorderWidth(10);
	final HBox hbox1 = new HBox(true, 0);
	final HBox hbox2 = new HBox(true, 0);
	final HBox hbox3 = new HBox(true, 0);

	final Label maxLabel = new Label(Translate._("Max. value:"));
	final Label minLabel = new Label(Translate._("Min. value:"));
	final Label lastLabel = new Label(Translate._("Last value:"));
	maxValue = new Label();
	minValue = new Label();
	lastValue = new Label();

	final Thread lastValueDaemon = new Thread(new ValueDaemon(minValue,
		maxValue, lastValue), "valueDaemon");
	lastValueDaemon.setDaemon(true);
	lastValueDaemon.start();

	hbox1.add(maxLabel);
	hbox1.add(maxValue);

	hbox2.add(minLabel);
	hbox2.add(minValue);

	hbox3.add(lastLabel);
	hbox3.add(lastValue);

	vbox.packStart(hbox1, false, false, 1);
	vbox.packStart(hbox2, false, false, 1);
	vbox.packStart(hbox3, false, false, 1);

	frame.add(vbox);
    }

    private void readConfig() {
	try {
	    final Config config = new Config();
	    fdSwitch.setActive(config.faceDetect());
	    llSpin.setValue(config.lowerLimit());
	    ulSpin.setValue(config.upperLimit());
	    miInSpin.setValue(config.minIncrease());
	    maInSpin.setValue(config.maxIncrease());
	    dtSpin.setValue(config.darkeningThreshold());
	    btSpin.setValue(config.brighteningThreshold());
	    loSpin.setValue(config.logLevel());
	} catch (final MissingResourceException e) {
	    Debug.LOG.log(Level.SEVERE, "Could not find config file.", e);
	} catch (final MalformedURLException e1) {
	    Debug.LOG.log(Level.SEVERE, "Could not read config file.", e1);
	}
    }

    private void constructMenu(MenuBar menuBar) {
	final MenuItem helpItem = new MenuItem(Translate._("_Help"));
	final MenuItem prefItem = new MenuItem(Translate._("_Preferences"));

	final Menu helpMenu = new Menu();
	final Menu prefMenu = new Menu();

	final MenuItem aboutItem = new MenuItem(Translate._("_About"));
	final AboutGUI aboutGUI = new AboutGUI(mainGUI, aboutItem);
	aboutGUI.start();
	final MenuItem faqItem = new MenuItem(Translate._("FAQ"));
	faqItem.connect(new Activate() {
	    @Override
	    public void onActivate(MenuItem arg0) {
		// Open launchpad.net project questions web page
		if (java.awt.Desktop.isDesktopSupported()) {
		    final java.awt.Desktop desktop = java.awt.Desktop
			    .getDesktop();

		    if (desktop.isSupported(java.awt.Desktop.Action.BROWSE)) {
			java.net.URI uri;
			try {
			    // Set URL for launchpad.net FAQ
			    uri = new java.net.URI(Database.QUESTIONS);
			    desktop.browse(uri);
			} catch (final URISyntaxException e) {
			    Debug.LOG.log(Level.SEVERE, "Wrong URL syntax: "
				    + Database.QUESTIONS, e);
			} catch (final IOException e) {
			    Debug.LOG.log(Level.WARNING,
				    "Could not start default browser.", e);
			}
		    }
		}
	    }
	});
	final MenuItem calibItem = new MenuItem(Translate._("_Calibrate"));
	final CalibGUI calibGUI = new CalibGUI(mainGUI, calibItem);
	calibGUI.start();

	helpMenu.add(aboutItem);
	helpMenu.add(faqItem);
	prefMenu.add(calibItem);

	helpItem.setSubmenu(helpMenu);
	prefItem.setSubmenu(prefMenu);

	menuBar.add(helpItem);
	menuBar.add(prefItem);
    }
}
